scylla/src/statement/prepared_statement.rs

```diff
@@ -337,3 +341,132 @@ pub enum PartitionKeyError {
+// meantime, this struct will return some garbage data while it tries to
+// decode the key, but nothing bad like a panic should happen otherwise.
+// The struct is currently only used for printing the partition key, so that's
+// completely fine.
+#[derive(Clone, Copy)]
+pub(crate) struct PartitionKeyDecoder<'pk> {
+    prepared_metadata: &'pk PreparedMetadata,
+    partition_key: &'pk [u8],
+    index: usize,
+}
+
+impl<'pk> PartitionKeyDecoder<'pk> {
+    pub(crate) fn new(prepared_metadata: &'pk PreparedMetadata, partition_key: &'pk [u8]) -> Self {
+        Self {
+            prepared_metadata,
+            partition_key,
+            index: 0,
+        }
+    }
+}
+
+impl<'pk> Iterator for PartitionKeyDecoder<'pk> {
+    type Item = Result<CqlValue, ParseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.prepared_metadata.pk_indexes.len() {
+            return None;
+        }
+
+        let idx = self.prepared_metadata.pk_indexes[self.index].index as usize;
+        let spec = &self.prepared_metadata.col_specs[idx];
+
+        let cell = if self.prepared_metadata.pk_indexes.len() == 1 {
+            Ok(self.partition_key)
+        } else {
+            self.parse_cell()
+        };
+        self.index += 1;
+        Some(cell.and_then(|mut cell| deser_cql_value(&spec.typ, &mut cell)))
+    }
+}
+
+impl<'pk> PartitionKeyDecoder<'pk> {
+    fn parse_cell(&mut self) -> Result<&'pk [u8], ParseError> {
+        let buf = &mut self.partition_key;
+        let len = buf.read_u16::<BigEndian>()? as usize;
+        if buf.len() < len {
+            return Err(ParseError::IoError(std::io::Error::new(
+                std::io::ErrorKind::UnexpectedEof,
+                "value too short",
+            )));
+        }
+        let col = &buf[..len];
+        *buf = &buf[len..];
+        buf.read_u8()?;
+        Ok(col)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bytes::{BufMut, Bytes, BytesMut};
+    use scylla_cql::frame::response::result::{
+        ColumnSpec, ColumnType, CqlValue, PartitionKeyIndex, PreparedMetadata, TableSpec,
+    };
+
+    use super::PartitionKeyDecoder;
+
+    fn make_pk(col_count: usize) -> (PreparedMetadata, Bytes) {
+        let table_spec = TableSpec {
+            ks_name: "ks".to_owned(),
+            table_name: "t".to_owned(),
+        };
+        let meta = PreparedMetadata {
+            flags: 0,
+            col_count,
+            col_specs: (0..col_count)
+                .map(|i| ColumnSpec {
+                    name: format!("col_{}", i),
+                    table_spec: table_spec.clone(),
+                    typ: ColumnType::Int,
+                })
+                .collect(),
+            pk_indexes: (0..col_count)
+                .map(|i| PartitionKeyIndex {
+                    index: i as u16,
+                    sequence: i as u16,
+                })
+                .collect(),
+        };
```

Comment:
Is such `PreparedMetadata` general enough to cover the most important test cases? I mean, column type is always `Int`, and sequence numbers are equal to indices. I'm not sure though if we need to test other cases _here_.